### PR TITLE
Ruby - Project 2 Enumerable methods

### PR DESCRIPTION
**Pull request by mariosknl**
This is Project 2 of Ruby Section in which I created my own methods based on the ones from the Enumerable module.

### CODE REVIEW

**Review by whiz25 — CHANGES_REQUESTED** [review]
## Nice Progress :+1: 
Nice try with the first attempt at the project. We have a lot to fix before we can merge this PR.
Some observations:
- [ ] enable stickler in this repo and make sure all tests pass
- [ ] check more comments on individual methods for suggestions

### Martin Afani

File: custom_enumerable.rb [hunk 1]
```diff
@@ -0,0 +1,116 @@
+
+# frozen_string_literal: true
+
+# rubocop:disable Metrics/PerceivedComplexity,Metrics/CyclomaticComplexity, Style/CaseEquality
+
+
+module Enumerable
+  def my_each
```
Comment (whiz25) [inline custom_enumerable.rb:3]: - [ ] should be identical and returns the same thing as ruby's [`each`](https://ruby-doc.org/core-2.6.4/Array.html#method-i-each). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] my_each returns an Enumerator if no block is given
```ruby
array = Array.new(100) { rand(0...9) } }
array.my_each) == Enumerator #1 true
```

File: custom_enumerable.rb [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+      i = 0
+      if block_given?
+      while i < self.length 
+          yield(self[i])
+          i += 1
+      end
+      end
+  end
+
+  def my_each_with_index
```
Comment (whiz25) [inline custom_enumerable.rb:14]: - [ ] should be identical and returns the same thing as ruby's [`each_with_index`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-each_with_index). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each_with_index.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] returns an enumerator if no block is given
```ruby
array = Array.new(100){rand(0...9)}
array.my_each_with_index == Enumerator #true
```

File: custom_enumerable.rb [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    i = 0
+    for i in i...self.length
+      yield(self[i], i)
+  end
+
+  def my_select
```
Comment (whiz25) [inline custom_enumerable.rb]: - [ ] should be identical and returns the same thing as ruby's [`select`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-select). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_select.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] returns an enumerator if no block is given
```ruby
array = Array.new(100){rand(0...9)}
array.my_select == Enumerator #true
```
- [ ] returns an array containing all elements of enum for which the given block returns a true value
```ruby
array = Array.new(100){rand(0...9)}
array.my_select(&block) == array.select(&block) # return true
```

File: custom_enumerable.rb [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    array = []
+    self.my_each { |i| array << array[i] if yield(i) }
+    end
+  end  
+
+  def my_all?
```
Comment (whiz25) [inline custom_enumerable.rb]: - [ ] should be identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] when no block or argument is given returns true when none of the collection members are false or nil
```ruby
true_array = [1, true, 'hi', []] 
false_array = [1, false, 'hi', []] 
true_array.my_all? == true_array.all? #true
```
- [ ] when a class is passed as an argument returns true if all of the collection is a member of such class
```ruby
array = Array.new(100){rand(0...9)}
array.my_all?(Integer) == array.all?(Integer)
```
- [ ] when a Regex is passed as an argument returns true if all of the collection matches the Regex
```ruby
words = %w[dog door rod blade]
words.my_all?(/d/) == words.all?(/d/)
```
- [ ] when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
```ruby
array.my_all?(3) == array.all?(3) #true
```
- [ ] returns true if the block never returns false or nil
```ruby
true_block = proc { |num| num <= 9 } 
false_block = proc { |num| num > 9 } 
array.my_all?(&true_block) == array.all?(&true_block) #true
array.my_all?(&false_block) == array.all?(&false_block) #true
```

File: custom_enumerable.rb [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    final = true
+    self.my_each { |i| true unless yield(i) }
+    final
+  end
+
+  def my_any?
```
Comment (whiz25) [inline custom_enumerable.rb]: - [ ] should be identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] when no block or argument is given returns true if at least one of the collection is not false or nil
```ruby
true_array = [nil, false, true, []]
true_array.my_any? == true_array.any? #true
```
- [ ]  when a class is passed as an argument returns true if at least one of the collection is a member of such class
```ruby
words = %w[dog door rod blade]
words.my_any?(Integer) == words.any?(Integer) #true
```
- [ ] when a Regex is passed as an argument returns false if none of the collection matches the Regex
```ruby
words.my_any?(/z/) == words.any?(/z/) #true
```
- [ ] when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
```ruby
words.my_any?('cat') == words.any?('cat') #true
```
- [ ] returns true if the block ever returns a value other than false or nil
```ruby
true_block = proc { |num| num <= 9 } 
false_block = proc { |num| num > 9 } 
array.my_any?(&true_block) == array.any?(&true_block) #true
array.my_any?(&false_block) == array.any?(&false_block) #true
```

File: custom_enumerable.rb [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    result = false
+    self.my_each { |i| false unless yield(i) }
+    result
+  end
+
+  def my_none?
```
Comment (whiz25) [inline custom_enumerable.rb]: - [ ] should be identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] when a class is passed as an argument returns true if none of the collection is a member of such class
```ruby
array = Array.new(100) { rand(0...9) }
array.my_none?(String) == true #true
```
- [ ] when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
```ruby
words = %w[dog door rod blade]
words.my_none?(5) == words.none?(5) #true
```
- [ ] Regex is passed as an argument returns true only if none of the collection matches the Regex
```ruby
 words.my_none?(/z/) == words.none?(/z/) #true
 words.my_none?(/d/) == words.none?(/d/) #true
```
- [ ] returns true if the block never returns true for all elements
```ruby
true_block = proc { |num| num > 9 }
false_block = proc { |num| num <= 9 } 

array.my_none?(&true_block) == array.none?(&true_block) #true
array.my_none?(&false_block) == array.none?(&false_block) #true
```
- [ ] when no block or argument is given
```ruby
true_array = [nil, false, true, []] 
false_array = [nil, false, nil, false]

false_array.my_none? == true #true
true_array.my_none? == false #false
```

File: custom_enumerable.rb [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+    result = true
+    self.my_each do |i|
+      result = false if self[i] == yield(i)
+    end
+    result
+  end
+
+  def my_count
```
Comment (whiz25) [inline custom_enumerable.rb]: - [ ] should be identical and returns the same thing as ruby's [`count`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-count). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_count.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] counts the number of items in enum that are equal to item if an argument is given
```ruby
array = Array.new(100){rand(0...9)}
array.my_count(0) == array.count(0) #true
```
- [ ] counts the number of elements yielding a true value if a block is given
```ruby
array.my_count(&block) == array.count(&block)
```

File: custom_enumerable.rb [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+    count = 0
+    self.my_each { |i| count += 1 }
+    count
+  end
+
+  def my_map(&my_proc)
```
Comment (whiz25) [inline custom_enumerable.rb]: - [ ] should be identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] returns a new array with the results of running block once for every element in enum.
```ruby
array = Array.new(100){rand(0...9)}
array.my_map(&block) == array.map(&block) #true
```
- [ ] returns an Enumerator if no block is given
```ruby
array.my_map == Enumerator #true
```

File: custom_enumerable.rb [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+
+  def my_map(&my_proc) 
+    final = []
+    my_each do |i| 
+      final << yield(i) if block_given? && yield(i) !=0 
+    end
+    print final 
+  end
+
+  def my_inject(acc = nil, operator = nil)
```
Comment (whiz25) [inline custom_enumerable.rb]: - [ ] should be identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] when a symbol is specified combines each element of the collection by applying the symbol as a named method
```ruby
array = Array.new(100) { rand(0...9) }
array.my_inject(:+) == array.inject(:+) #true
```
- [ ] combines all elements of enum by applying a binary operation, specified by a block
```ruby
operation = proc { |sum, n| sum + n } 
search = proc { |memo, word| memo.length > word.length ? memo : word } 
array = Array.new(100){rand(0...9)}
words = %w[dog door rod blade]

array.my_inject(&operation) == array.inject(&operation) #true
actual = range.my_inject(4) { |prod, n| prod * n }
expected = range.inject(4) { |prod, n| prod * n }
actual == expected #true
words.my_inject(&search) == words.inject(&search) #true
```

**Review by euqueme — CHANGES_REQUESTED** [review]
> ### Pull Request under the review by Maria Eugenia Queme P. (TSE)

### Hello @mariosknl  :handshake: 

### Thank you for doing some of the changes required by the previous reviewer, :medal_sports:  your project looks better but we still have to work on some of your enumerable methods :construction_worker_man: :construction: 

## my_each_with_index

- [x] calls the given block once for each element in self

Your my_each_with_index is not operating the last element of the Enumerable, let's fix it

- Yours

image

- Real

image

## my_select

- [x] returns an array containing all elements of enum for which the given block returns a true value

Your my_select has to return an array containing all the elements that met the condition given in the block, right now is just returning nil

Given

  > block = proc { |num| num < (0 + 9) / 2 } 

- Yours
image

- Real
image

## my_all?

- [x] when a class is passed as an argument returns true only if all of the collection is a member of such class, otherwise, it should return false

Your my_all is returning true even though not all elements met the condition given in the argument

- Yours
image

- Real
image

- [ ] when a Regex is passed as an argument returns true if all of the collection matches the Regex

- Yours
image

- Real
image

- [ ] when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern

- Yours
image

- Real
image

## my_any?

- [x] when a class is passed as an argument returns true if at least one of the collection is a member of such class

 Your my_any? displays an error when passing a class as an argument, let's fix it

- Yours
image

- Real
image

## my_none?

- [x] when a class is passed as an argument returns true if none of the collection is a member of such class

 Your my_none? displays an error when passing a class as an argument, let's fix it

- Yours
image

- Real
image

[OPTIONAL]

I think if we could get the block of my_none? from my_any? we should only do something like !my_any? for my_none? to work. For example:

```
  def my_none?(arg = nil, &block)
    !my_any?(arg = nil,&block)
  end
```

To make this work we have to use block.call(element) instead of yield(element) in my_any?

## my_inject

- [ ] combines all elements of enum by applying a binary operation, specified by a block

Yours is giving an error when both and argument and a block is passed

- Yours
image

- Real
image

- [ ] When a symbol is provided as an argument, my_inject should combine the elements of the enum using that symbol as a named method.

Yours is giving the wrong answer for some test cases

Given

> array = Array.new(100) { rand(0...9) }

- Yours
image

- Real
image

## my_map

[OPTIONAL]

- [ ] in the specifications of the project in the [Odin project](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks) # 12, you were asked to modify your my_map method to take either a proc or a block. And if both a proc and a block are given, only execute the proc. The real map doesn't have this feature, but you were asked to have it so I strongly recommend to make it work.

Given 
`my_proc = Proc.new {|num| num > 10 }`

Result
> puts [18, 22, 5, 6] .my_map(my_proc) {|num| num < 10 }  # true true false false

## Rubocop errors

- [ ] your project is showing rubocop errors, please fix them before submitting for a code review again, your last commit has to have a green check mark :heavy_check_mark: 

### That's it! :raised_hands: 

### Please :pray: do the changes :man_technologist: and submit back for a code review when you are done :memo: 

Happy coding! :rocket: 

Any question? ask me on slack **@ Maru** :woman_technologist: 

PROJECT STATUS: REQUESTED CHANGES :traffic_light:

File: custom_enumerable.rb [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+        i += 1
+      end
+    end
+  end
+
+  def my_select(*)
+    return to_enum(:my_select) unless block_given?
+
+    array = []
+    my_each { |i| p array << i if yield(i) }
```
Comment (MauricioRobayo) [inline custom_enumerable.rb]: Seems like you are printing the array 🤔, and what are you returning?

**Review by MauricioRobayo — CHANGES_REQUESTED** [review]
Very nice project!

It looks great and is coming into place, but we need to fix some issues before continuing:


#### `my_select`

* [ ] returns an array containing all elements of enum for which the given block returns a true value

```ruby
>> [1, 2, 3].select { |n| n == 2 }
=> [2]
>> [1, 2, 3].my_select { |n| n == 2 }
[2]
=> nil
```

#### `my_all?`

* [ ] returns true if the block never returns false or nil
```ruby
>> [1].all? { |n| n == 1 }
=> true
>> [1].my_all? { |n| n == 1 }
=> false
```

* [ ] when no block or argument is given returns true when none of the collection members are false or nil

```ruby
>> [1].all?
=> true
>> [1].my_all?
=> #<Enumerator: [1]:my_select>
```

* [ ] when a class is passed as an argument returns true if all of the collection is a member of such class
```ruby
>> [1].all?(Integer)
=> true
>> [1].my_all?(Integer)
=> #<Enumerator: [1]:my_select>
```

* [ ] when a Regex is passed as an argument returns true if all of the collection matches the Regex
```ruby
>> %w[a b].all?(/\w/)
=> true
>> %w[a b].my_all?(/\w/)
=> #<Enumerator: ["a", "b"]:my_select>
```

* [ ] when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
```ruby
>> [1].all?(1)
=> true
>> [1].my_all?(1)
=> #<Enumerator: [1]:my_select>
```

Status: __Request changes__.

---

Optional:

* [ ] In the [Odin project #12](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2) you were asked to modify your `my_map` method to take either a proc or a block. And if both a proc and a block are given, only execute the proc. The real map doesn't have this feature.

```ruby
>> my_proc = Proc.new { |num| num > 5 }
>> [4, 6].my_map(my_proc) # should return false true
>> [4, 6].my_map(my_proc) { |num| num < 5 } # should return false true
```

**Review by whiz25 — CHANGES_REQUESTED** [review]
## Good Job @mariosknl :+1: 
You are one step away from merging this PR. Below is the only failing edge case.
- [ ] should be identical and returns the same thing as ruby's [`select`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-select). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_select.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] returns an array containing all elements of enum for which the given block returns a true value
```ruby
array = Array.new(100){rand(0...9)}
array.my_select(&block) == array.select(&block) # return true
```

### Martin Afani

**Review by whiz25 — APPROVED** [review]
## Congz :clap: 
All requested changes have been made and it's a wrap for this PR. Merge and proceed.

### Martin Afani

**Pull request merged**